### PR TITLE
GDPR-160 Correcting bug where we weren't recording all offender ids.

### PR DESCRIPTION
**Pull request by brightonsbox**
Any offender ids that weren't linked to a booking were getting
missed. Spotted the issue because now we echo back the ids
at point of deletion to ensure no new aliases / bookings
have been created since the initial referral.

### CODE REVIEW

File: src/main/java/uk/gov/justice/hmpps/datacompliance/repository/jpa/model/referral/ReferredOffenderIds.java [hunk 1]
```diff
@@ -43,6 +43,6 @@
     private Long offenderId;
 
     @NotNull
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/datacompliance/repository/jpa/model/referral/ReferredOffenderIds.java:45]: still got `NotNull` here?

File: src/main/java/uk/gov/justice/hmpps/datacompliance/repository/jpa/model/referral/ReferredOffenderIds.java [hunk 2]
```diff
@@ -26,8 +26,8 @@
 @AllArgsConstructor
 @EqualsAndHashCode(of = {"referredOffenderBookingId"})
 @ToString(exclude = "offenderDeletionReferral") // to avoid circular reference
-@Table(name = "REFERRED_OFFENDER_BOOKING")
-public class ReferredOffenderBooking {
+@Table(name = "REFERRED_OFFENDER_IDS")
+public class ReferredOffenderIds {
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/datacompliance/repository/jpa/model/referral/ReferredOffenderIds.java:30]: why is this plural?  Isn't it just holding a single link to an offender and booking?

File: src/main/java/uk/gov/justice/hmpps/datacompliance/repository/jpa/model/referral/OffenderDeletionReferral.java [hunk 3]
```diff
@@ -91,13 +91,13 @@ public OffenderDeletionReferral(final Long referralId,
...
     @OneToMany(mappedBy = "offenderDeletionReferral", cascade = PERSIST, fetch = LAZY)
-    private final List<ReferredOffenderBooking> offenderBookings = new ArrayList<>();
+    private final List<ReferredOffenderIds> offenderIds = new ArrayList<>();
 
     @OneToOne(mappedBy = "offenderDeletionReferral", cascade = PERSIST, fetch = LAZY)
     private ReferralResolution referralResolution;
 
-    public OffenderDeletionReferral addReferredOffenderBooking(final ReferredOffenderBooking booking) {
-        this.offenderBookings.add(booking);
+    public OffenderDeletionReferral addReferredOffenderIds(final ReferredOffenderIds booking) {
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/datacompliance/repository/jpa/model/referral/OffenderDeletionReferral.java:99]: This code is quite confusing.  Your passing in a singular referred offender IDs that you're calling a booking to then add to the list of offender IDs that isn't a list a offender IDs.  A backref is then set on the booking and the deletion referral is returned even though it is never used?

Would be better to do something like
```java
    public void addReferredOffenderId(final ReferredOffenderId referredOffenderId) {
        referredOffenderIds.add(referredOffenderId);
        referredOffenderId.setOffenderDeletionReferral(this);
    }
```
although I'm not happy about that either.  `ReferredOffenderId` sounds like it is the offender id of a referral but it isn't.  Maybe `OffenderReferral` instead?

File: src/main/java/uk/gov/justice/hmpps/datacompliance/repository/jpa/model/referral/ReferredOffenderIds.java (hunk 2 above)
Comment (petergphillips) [reply src/main/java/uk/gov/justice/hmpps/datacompliance/repository/jpa/model/referral/ReferredOffenderIds.java:30]: It reads like it holds a list of referred offender IDs but each row in the table is an offender record with an optional booking record

Comment (petergphillips) [reply src/main/java/uk/gov/justice/hmpps/datacompliance/repository/jpa/model/referral/ReferredOffenderIds.java:30]: The `REFERRED_OFFENDER_BOOKING_ID` column sounds like it is the offender booking id relating to the referral, but should really be renamed to match the new table name

File: src/main/java/uk/gov/justice/hmpps/datacompliance/services/referral/ReferralService.java [hunk 4]
```diff
@@ -68,12 +68,14 @@ private OffenderDeletionReferral createReferral(final OffenderPendingDeletion ev
...
-                offender.getOffenderBookings().forEach(booking ->
-                        referral.addReferredOffenderBooking(ReferredOffenderBooking.builder()
-                                .offenderId(offender.getOffenderId())
-                                .offenderBookId(booking.getOffenderBookId())
-                                .build())));
+        event.getOffenders().forEach(offender -> {
+
+            final var offenderIds = ReferredOffenderIds.builder().offenderId(offender.getOffenderId());
+
+            offender.getOffenderBookings().forEach(booking -> offenderIds.offenderBookId(booking.getOffenderBookId()));
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/datacompliance/services/referral/ReferralService.java:75]: not understanding this code.  We're looping over each offender relating to an event.  For each one we're constructing a singular referred offender (variable is `offenderIds` though).  For each offender booking we're then setting the offender book id of the referred offender.  Won't this overwrite each time - why aren't you just calling `last` on the offender bookings and set the referred offending book id to that one - why the loop over the offender bookings?

File: src/test/java/uk/gov/justice/hmpps/datacompliance/services/referral/ReferralServiceTest.java [hunk 5]
```diff
@@ -116,9 +116,10 @@ private OffenderPendingDeletion generatePendingDeletionEvent() {
                 .lastName("Smith")
                 .birthDate(LocalDate.of(1969, 1, 1))
                 .offender(OffenderWithBookings.builder()
-                        .offenderId(1L)
-                        .offenderBooking(OffenderBooking.builder().offenderBookId(2L).build())
+                        .offenderId(123L)
+                        .offenderBooking(OffenderBooking.builder().offenderBookId(321L).build())
```
Comment (petergphillips) [inline src/test/java/uk/gov/justice/hmpps/datacompliance/services/referral/ReferralServiceTest.java:120]: would be good to have multiple bookings in the test here

File: src/main/java/uk/gov/justice/hmpps/datacompliance/repository/jpa/model/referral/ReferredOffenderIds.java (hunk 2 above)
Comment (brightonsbox) [reply src/main/java/uk/gov/justice/hmpps/datacompliance/repository/jpa/model/referral/ReferredOffenderIds.java]: I've renamed to `ReferredOffenderAlias`.  Really I probably should have had a separate table just to store the list of booking ids for a given offender id to keep normal form, but that seemed a bit overkill at the time just to store one id.

File: src/main/java/uk/gov/justice/hmpps/datacompliance/repository/jpa/model/referral/OffenderDeletionReferral.java (hunk 3 above)
Comment (brightonsbox) [reply src/main/java/uk/gov/justice/hmpps/datacompliance/repository/jpa/model/referral/OffenderDeletionReferral.java]: Agreed, it didn't help that I hadn't properly renamed everything, sorry!!  I'm now using `ReferredOffenderAlias`

File: src/main/java/uk/gov/justice/hmpps/datacompliance/services/referral/ReferralService.java (hunk 4 above)
Comment (brightonsbox) [reply src/main/java/uk/gov/justice/hmpps/datacompliance/services/referral/ReferralService.java]: Totally agree, this wouldn't have worked, I've corrected the logic now.

**Pull request merged**